src/unix/aio.c

```diff
@@ -0,0 +1,352 @@
+{
+    if (!iocb) {
+        return -EFAULT;
+    }
+    thread_log(current, "%s: fd %d, op %d", __func__, iocb->aio_fildes,
+            iocb->aio_lio_opcode);
+
+    fdesc f = resolve_fd(current->p, iocb->aio_fildes);
+    if (aio->ongoing_ops >= aio_avail_events(aio) - 1) {
+        return -EAGAIN;
+    }
+    if (iocb->aio_reserved1 || iocb->aio_reserved2 || !iocb->aio_buf ||
+            (iocb->aio_flags & ~AIO_KNOWN_FLAGS)) {
+        return -EINVAL;
+    }
+
+    int res_fd;
+    if (iocb->aio_flags & IOCB_FLAG_RESFD) {
+        res_fd = iocb->aio_resfd;
+    } else {
+        res_fd = AIO_RESFD_INVALID;
+    }
+    io_completion completion = closure(heap_general(aio->kh), aio_complete, aio,
+            iocb->aio_data, (u64) iocb, res_fd);
+    switch (iocb->aio_lio_opcode) {
+    case IOCB_CMD_PREAD:
+        if (!f->read) {
+            goto inval;
+        }
+        apply(f->read, (void *) iocb->aio_buf, iocb->aio_nbytes,
+                iocb->aio_offset, current, true, completion);
+        break;
+    case IOCB_CMD_PWRITE:
+        if (!f->write) {
+            goto inval;
+        }
+        apply(f->write, (void *) iocb->aio_buf, iocb->aio_nbytes,
+                iocb->aio_offset, current, true, completion);
+        break;
+    default:
+        goto inval;
+    }
+    aio_lock(aio);
+    aio->ongoing_ops++;
+    aio_unlock(aio);
+    return 0;
+inval:
+    deallocate_closure(completion);
+    return -EINVAL;
+}
+
+sysreturn io_submit(aio_context_t ctx_id, long nr, struct iocb **iocbpp)
+{
+    struct aio *aio;
+    if (!ctx_id) {
+        return -EFAULT;
+    }
+    if (!(aio = aio_from_ring(current->p, ctx_id))) {
+        return -EINVAL;
+    }
+    int io_ops;
+    for (io_ops = 0; io_ops < nr; io_ops++) {
+        sysreturn rv = iocb_enqueue(aio, iocbpp[io_ops]);
+        if (rv) {
+            if (io_ops == 0) {
+                return rv;
+            } else {
+                break;
+            }
+        }
+    }
+    return io_ops;
+}
+
+closure_function(5, 1, sysreturn, io_getevents_bh,
+                 struct aio *, aio, long, min_nr, long, nr, struct io_event *, events, thread, t,
+                 u64, flags)
+{
+    struct aio *aio = bound(aio);
+    struct io_event *events = bound(events);
+    thread t = bound(t);
+    aio_ring ring = aio->ring;
+    aio_lock(aio);
+    unsigned int head = ring->head;
+    unsigned int tail = ring->tail;
+    if (head >= aio->nr) {
+        head = 0;
+    }
+    if (tail >= aio->nr) {
+        tail = 0;
+    }
+    while (head != tail) {
+        runtime_memcpy(&events[aio->copied_evts], &ring->events[head],
+                sizeof(struct io_event));
+        if (++head == aio->nr) {
+            head = 0;
+        }
+        if (++aio->copied_evts == bound(nr)) {
+            break;
+        }
+    }
+    ring->head = head;
+    ring->tail = tail;
+    aio_unlock(aio);
+    sysreturn rv = 0;
+    if (flags & BLOCKQ_ACTION_NULLIFY) {
```

Comment:
Maybe nullify should be checked prior to event dequeue? Usually a nullify means that the blocking syscall was interrupted by a signal, effectively becoming a nop and thus requiring the syscall to be later reissued. Although I'm not sure how this plays out for aio.

Reply:
Right, a signal can interrupt io_getevents() just like it does for other syscalls, so if a nullify action is delivered we shouldn't bother checking whether there are events to be dequeued. Fixed.